stores/sqlstore.go

```diff
@@ -359,14 +391,156 @@ func NewSQLStore(log logger.Logger, driver, source string, limits *StoreLimits,
+		s.dbLock = &sqlDBLock{
+			id: nuid.New().Next(),
+			db: s.db,
+		}
+	}
+	if s.dbLock.isOwner {
+		return true, nil
+	}
+	hasLock, id, tick, err := s.acquireDBLock(false)
+	if err != nil {
+		return false, err
+	}
+	if !hasLock {
+		// We did not get the lock. Try to see if the table is updated
+		// after 1 interval. If so, consider the lock "healthy" and just
+		// return that we did not get the lock. If after a configured
+		// number of tries the tick for current owner is not updated,
+		// steal the lock.
+		prevID := id
+		prevTick := tick
+		for i := 0; i < sqlLockLostCount; i++ {
+			time.Sleep(time.Duration(1.5 * float64(sqlLockUpdateInterval)))
+			hasLock, id, tick, err = s.acquireDBLock(false)
+			if hasLock || err != nil || id != prevID || tick != prevTick {
+				return hasLock, err
+			}
+			prevTick = tick
+		}
+		// Try to steal.
+		hasLock, _, _, err = s.acquireDBLock(true)
+	}
+	if hasLock {
+		// Success. Keep track that we own the lock so we can clear
+		// the table on clean shutdown to release the lock immediately.
+		s.dbLock.Lock()
+		s.dbLock.isOwner = true
+		s.wg.Add(1)
+		go s.updateDBLock()
+		s.dbLock.Unlock()
+	}
+	return hasLock, err
+}
+
+// This go-routine updates the DB store lock at regular intervals.
+func (s *SQLStore) updateDBLock() {
+	defer s.wg.Done()
+
+	var (
+		ticker  = time.NewTicker(sqlLockUpdateInterval)
+		hasLock = true
+		err     error
+		failed  int
+	)
+	for {
+		select {
+		case <-ticker.C:
+			hasLock, _, _, err = s.acquireDBLock(false)
+			if !hasLock || err != nil {
+				// If there is no error but we did not get the lock,
+				// something is really wrong, abort right away.
+				stopNow := !hasLock && err == nil
+				if err != nil {
+					failed++
+					s.log.Errorf("Unable to update store lock (failed=%v err=%v)", failed, err)
+				}
+				if stopNow || failed == sqlLockLostCount {
+					if sqlNoPanic {
+						s.log.Fatalf("Aborting")
+						return
+					}
+					panic("lost store lock, aborting")
+				}
+			} else {
+				failed = 0
+			}
+		case <-s.doneCh:
+			ticker.Stop()
+			return
+		}
+	}
+}
+
+// Returns if lock is acquired, the owner and tick value of the lock record.
+func (s *SQLStore) acquireDBLock(steal bool) (bool, string, uint64, error) {
+	s.dbLock.Lock()
+	defer s.dbLock.Unlock()
+	var (
+		lockID  string
+		tick    uint64
+		hasLock bool
+	)
+	tx, err := s.dbLock.db.Begin()
+	if err != nil {
+		return false, "", 0, err
+	}
+	defer func() {
+		if tx != nil {
+			tx.Rollback()
+		}
+	}()
+	r := tx.QueryRow(sqlStmts[sqlDBLockSelect])
+	err = r.Scan(&lockID, &tick)
+	if err != nil && err != sql.ErrNoRows {
+		return false, "", 0, sqlStmtError(sqlDBLockSelect, err)
+	}
+	if err == sql.ErrNoRows || steal || lockID == "" || lockID == s.dbLock.id {
+		// If we are stealing, reset tick to 0 (so it will become 1 in update statement)
+		if steal {
+			tick = 0
+		}
+		stmt := sqlStmts[sqlDBLockUpdate]
+		if err == sql.ErrNoRows {
+			stmt = sqlStmts[sqlDBLockInsert]
```

Comment:
I really want to believe this doesn't work this way, but tested it directly in MySQL and it does! Thought `FOR UPDATE` only locked selected rows.

Reply:
Yes, I first I required to have at least a row in that table (even when creating the table, I had an insert with blank values in there). But then I realized that it would work in that if there is 2 selects for update when the table is empty, one of them will fail at the time of commit, so that solves the issue.